pkg/tfshim/sdk-v2/proposed_new.go

```diff
@@ -0,0 +1,200 @@
+	hctyjson "github.com/hashicorp/go-cty/cty/json"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/zclconf/go-cty/cty"
+	ctyjson "github.com/zclconf/go-cty/cty/json"
+
+	"github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfshim/sdk-v2/internal/tf/configs/configschema"
+	"github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfshim/sdk-v2/internal/tf/plans/objchange"
+)
+
+func proposedNew(res *schema.Resource, prior, config hcty.Value) (hcty.Value, error) {
+	schema, err := configschemaBlock(res)
+	if err != nil {
+		return hcty.False, err
+	}
+	priorC, err := hcty2cty(prior)
+	if err != nil {
+		return hcty.False, err
+	}
+	configC, err := hcty2cty(config)
+	if err != nil {
+		return hcty.False, err
+	}
+	return cty2hcty(objchange.ProposedNew(schema, priorC, configC))
+}
+
+func htype2ctype(t hcty.Type) (cty.Type, error) {
+	if t.Equals(hcty.NilType) {
+		return cty.NilType, nil
+	}
+	typeJSON, err := hctyjson.MarshalType(t)
+	if err != nil {
+		return cty.Bool, err
+	}
+	return ctyjson.UnmarshalType(typeJSON)
+}
+
+func hcty2cty(v hcty.Value) (cty.Value, error) {
+	if v.Equals(hcty.NilVal).True() {
+		return cty.NilVal, nil
+	}
+	typ, err := htype2ctype(v.Type())
+	if err != nil {
+		return cty.False, err
+	}
+	valueJSON, err := hctyjson.Marshal(v, v.Type())
+	if err != nil {
+		return cty.False, err
+	}
+	return ctyjson.Unmarshal(valueJSON, typ)
+}
+
+func cty2hcty(v cty.Value) (hcty.Value, error) {
+	if v.Equals(cty.NilVal).True() {
+		return hcty.NilVal, nil
+	}
+
+	typeJSON, err := ctyjson.MarshalType(v.Type())
+	if err != nil {
+		return hcty.False, err
+	}
+	valueJSON, err := ctyjson.Marshal(v, v.Type())
+	if err != nil {
+		return hcty.False, err
+	}
+	typ, err := hctyjson.UnmarshalType(typeJSON)
+	if err != nil {
+		return hcty.False, err
+	}
+	return hctyjson.Unmarshal(valueJSON, typ)
+}
+
+func configschemaBlock(res *schema.Resource) (*configschema.Block, error) {
+	schema := res.CoreConfigSchema()
+
+	block := &configschema.Block{
+		Attributes:      map[string]*configschema.Attribute{},
+		BlockTypes:      map[string]*configschema.NestedBlock{},
+		Description:     schema.Description,
+		DescriptionKind: configschema.StringKind(int(schema.DescriptionKind)),
+		Deprecated:      schema.Deprecated,
+	}
+
+	for name, a := range schema.Attributes {
+		t, err := htype2ctype(a.Type)
+		if err != nil {
+			return nil, err
+		}
+		block.Attributes[name] = &configschema.Attribute{
+			Type:            t,
+			Description:     a.Description,
+			DescriptionKind: configschema.StringKind(int(schema.DescriptionKind)),
+			Required:        a.Required,
+			Optional:        a.Optional,
+			Computed:        a.Computed,
+			Sensitive:       a.Sensitive,
+			Deprecated:      a.Deprecated,
+		}
+	}
+
+	queue := newQueue(schema.BlockTypes)
+
+	destinations := map[interface{}]map[string]*configschema.NestedBlock{}
+	for _, b := range schema.BlockTypes {
+		destinations[b] = block.BlockTypes
+	}
+
+	for !queue.empty() {
```

Comment:
It would be really helpful to have a comment describing what this algorithm is doing.

Reply:
Ah yes this one is quite an embarrassment. I have `a->b` function that' recursive but since type a is an internal type from a foreign go module I can't seem to be able to write out `func (x a) b` because I can't reference `a`! This does not prevent the program from manipulating values of type `a` and in fact storing them in a queue :) The rest are workarounds. Let me try a few more direct workarounds using generics, if not I'll just add a comment.

Reply:
Added a comment.